platform-sdk/docs/proposals/consensus-layer/Consensus-Layer.md

```diff
@@ -0,0 +1,438 @@
+whatever algorithm it chooses).
+
+#### Events
+
+Gossip is event-oriented, meaning that it is given events to gossip, and emits events it receives through gossip. An
+implementation of Gossip could be based on a lower-level implementation based on bytes, but at the module level, it
+works in terms of events.
+
+When the Gossip module receives events through gossip, it *may* choose to perform some deduplication before sending
+them to Event Intake, but it is not required to do so.
+
+Honest nodes will **only** send valid events in *topological order* to peers. A peer may still receive events out of
+order, because different events may arrive from different peers at different times. To support topological ordering,
+events received by Gossip are not retained by Gossip, they are simply passed directly to Event Intake, which after
+verification, also orders events in topological order and recorded durably to disk, and then send back to the Gossip
+module to be sent out to its gossip peers.
+
+**The key design principle here is that a node will only send valid events through gossip**. If invalid events are
+ever received, you may know the node that sent them to you is dishonest. Validating events increases latency at each
+"hop", but allows us to identify dishonest gossip peers and discipline them accordingly.
+
+During execution, for all nodes that are online and able to keep up, events are received "live" and processed
+immediately and re-gossiped. However, if a node is offline and then comes back online, or is starting back up after
+reconnect, it may be missing events. In this case, the node will need to ask its peers for any events it is missing.
+
+For this reason, every honest node needs to buffer some events, so when its peer asks it for events, it is able to
+give them. Events can be "non-ancient", "ancient", or "expired". Although the definition for these is subject to
+configuration, we will here use some example values. Non-ancient events are those events that are either not yet
+in a consensus round, or are in one of the last 26 consensus rounds. Expired events are those that are in a round
+that is older than the last most recent 500 consensus rounds. And ancient events are those that are in a round between
+26 and 500.
+
+The Gossip module must cache all non-expired events. Note that it must be a fixed number of rounds that defines
+"expired", because otherwise it is possible for a quiescent network to never recover (see the documentation for
+the tipset algorithm, and other-parent selection, for specifics on why this is true). This number must also be large
+enough in case rounds/sec increases dramatically, because reconnect requires a certain amount of time (not rounds) for
+catchup. As long as this number is large enough to meet th requirements regardless of the rounds/sec, and as long as it
+is fixed in number of rounds, then it will work.
```

Comment:
See comment above, number of rounds worth of events is a little fuzzy. Sure, we have approximately 500 rounds worth of events. But the exact boundary condition probably isn't accurately described by a "fixed number of rounds".

Reply:
What I really wanted to do was change this from being a round-based number to a time-based number. The problem with time-based, is if the network quiesces, then comes back, if it comes back well after that time based number, then there wouldn't be any events to gossip and "wake up" the tipset algo.

I agree it doesn't need to be fixed, and actually fixed is problematic because for reconnect you really do want this to be time-based (i.e. I have 8 minutes to "catch up") whereas if rounds/sec were to change dramatically based on network behavior, then maybe now I only have 2 minutes to "catch up".